### PR TITLE
Elastic search siem enhanchment

### PR DESCRIPTION
**Pull request by iyeshaya**
Enhance Elasticsearch_V2 to support EQL

### CODE REVIEW

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 1]
```diff
@@ -48,13 +48,18 @@
...
 TIME_METHOD = param.get('time_method', 'Simple-Date')
 TIMEOUT = int(param.get('timeout') or 60)
 MAP_LABELS = param.get('map_labels', True)
+TIME_RANGE_START = param.get('time_range_start')
+TIME_RANGE_END = param.get('time_range_end')
+
+FETCH_QUERY = RAW_QUERY or FETCH_QUERY_PARM
 
 
 def get_timestamp_first_fetch(last_fetch):
```
Comment (ShahafBenYakir) [inline Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py:65]: Let's rename this function as the usability has changed and we use this for several time-casting

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 2]
```diff
@@ -590,36 +606,54 @@ def format_to_iso(date_string):
...
-    # handle first time fetch
-    if last_fetch is None:
-        last_fetch, _ = parse_date_range(date_range=FETCH_TIME, date_format='%Y-%m-%dT%H:%M:%S.%f', utc=False,
-                                         to_timestamp=False)
-        last_fetch = parse(str(last_fetch))
-        last_fetch_timestamp = int(last_fetch.timestamp() * 1000)
+def get_time_range(last_fetch: Union[str, None] = None, fetch_time=FETCH_TIME, time_range_start=TIME_RANGE_START,
+                   time_range_end=TIME_RANGE_END, time_field=TIME_FIELD) -> Dict:
+    """
+    Creates the time range filter's dictionary based on the last fetch and given params.
```
Comment (ShahafBenYakir) [inline Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py:612]: Lets try to give more details on what is the logic we work by to create the time filter, which may help in debugging/solving bugs later on.

**Review by ShahafBenYakir — COMMENTED** [review]
Nice
1. Need to set the classifier and mapper by default in the YML.
1. Let's add the new commands to test  PB and the unit test file.
1. I think we are missing the permissions section in the integration readme.
1. Pls set up a 30 mins demo including TPM

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md [hunk 3]
```diff
@@ -1,339 +1,178 @@
...
+
+`es-search`
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| index | The index in which to perform a search. | Required | 
+| query | The string to query (in Lucene syntax). Possible values are: . | Required | 
+| fields | A comma-separated list of document fields to fetch. If empty, the entire document is fetched. | Optional | 
+| explain | Calculates an explanation of a score for a query. Default is "false". For example, "value:1.6943597". Possible values are: true, false. Default is false. | Optional |
```
[the remark this answers is not in the document]
Comment (ShahafBenYakir) [reply Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md:55]: I think we still have duplicate "default" mentioning in the README file

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py [hunk 4]
```diff
@@ -590,38 +629,75 @@ def format_to_iso(date_string):
...
+    """
+    range_dict = {}
+    if not last_fetch:  # this is the first fetch
+        # if time range start provided we will use it instead of fetch time.
+        if time_range_start:
+            start_date = dateparser.parse(time_range_start)
+        elif fetch_time:
+            start_date = dateparser.parse(fetch_time)
+        else:
+            raise DemistoException("Missing First fetch timestamp or Time Range Start, please provide one of them.")
```
Comment (ShahafBenYakir) [inline Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.py:659]: from going over this function code I don't see why we need these two arguments, we can just do the following:
```python
if not last_fetch: # this is the first fetch
    start_date = dateparser.parse(time_range_start)
```
In the `fetch_incidents` you will do:
```python
get_time_range(time_range_start=last_fetch)
```
And `search_command` you will do:
```python
get_time_range(time_range_start=time_range_start)
```

**Review by ShahafBenYakir — COMMENTED** [review]
Good. I see we have some leftovers from previous CR:
- Sections 1-3 from the previous general review comment.
- Some code comments are still not fixed. I left them unresolved, the rest I resolved.

**Comment by ShirleyDenkberg** [comment]
@iyeshaya 
In the Elasticsearch PR https://github.com/demisto/content/pull/20310, I found a discrepancy between the parameters in the readme and yml.
In the yml, I see Query String but in the readme, I see
 | Query by which to fetch incidents (Lucene syntax)
| Query Time Range - Start |
  | Query Time Range - End |
Also, in the Readme, I see a Password parameter, but it does not appear in the yml file.

**Comment by ShirleyDenkberg** [comment]
@iyeshaya Need to create a pack readme for this pack.
I started putting something together but I need more information.
Basically, the information I need is:
- What is a real life situation that would require me to need this pack?
- What does this pack do?
- Why do I want to download this pack?

This is what I have so far. Please provide me with additional information. 

> No matter where your data is located (in a single cloud, across multiple clouds, or on-premises) you can use the Elasticsearch content pack to search and analyze your data in real time.
> 
> ## What does this pack do?
> - Query and search indexes.
> - Calculate query and search scores based on accuracy of results.

**Comment by iyeshaya** [comment]
> @iyeshaya Need to create a pack readme for this pack. I started putting something together but I need more information. Basically, the information I need is:
> 
> * What is a real life situation that would require me to need this pack?
> * What does this pack do?
> * Why do I want to download this pack?
> 
> This is what I have so far. Please provide me with additional information.
> 
> > No matter where your data is located (in a single cloud, across multiple clouds, or on-premises) you can use the Elasticsearch content pack to search and analyze your data in real time.
> > ## What does this pack do?
> > 
> > * Query and search indexes.
> > * Calculate query and search scores based on accuracy of results.

This integration allows you to query elasticsearch instances using dsl, eql and lucene syntaxes.  In addition you can fetch incidents with predefined query. 

@bakatzir Can you help with the pack description?

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.yml [hunk 5]
```diff
@@ -156,6 +178,24 @@ script:
       - desc
       required: false
       secret: false
+    - default: false
+      description: Used instead of ‘query'.
+      isArray: false
+      name: query_dsl
+      required: false
+      secret: false
```
[the remark this answers is not in the document]
Comment (iyeshaya) [reply Packs/Elasticsearch/Integrations/Elasticsearch_v2/Elasticsearch_v2.yml]: @danikdanik  Should we change the name from query_dsl to raw_query?

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md [hunk 6]
```diff
@@ -1,339 +1,178 @@
...
-<li>Create a new mapper and edit an existing one</li>
-<li>Set <strong>Get data</strong> to <strong>Select schema</strong> and <strong>Select instance</strong> to the Elasticsearch v2 integration instance to fetch from.</li>
-<li>Map fields according to the fetched index schema.</li>
-</ol>
+Search for and analyze data in real time. 
+ Supports version 6 and later.
+This integration was integrated and tested with version 8.3 of Elasticsearch v2
+
+Some changes have been made that might affect your existing content. 
+If you are upgrading from a previous of this integration, see [Breaking Changes](#breaking-changes-from-the-previous-version-of-this-integration-elasticsearch-v2).
```
[the remark this answers is not in the document]
Comment (iyeshaya) [reply Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md:6]: This was added automatically. fixed.

**Comment by bakatzir** [comment]
> > @iyeshaya Need to create a pack readme for this pack. I started putting something together but I need more information. Basically, the information I need is:
> > 
> > * What is a real life situation that would require me to need this pack?
> > * What does this pack do?
> > * Why do I want to download this pack?
> > 
> > This is what I have so far. Please provide me with additional information.
> > > No matter where your data is located (in a single cloud, across multiple clouds, or on-premises) you can use the Elasticsearch content pack to search and analyze your data in real time.
> > > ## What does this pack do?
> > > 
> > > * Query and search indexes.
> > > * Calculate query and search scores based on accuracy of results.
> 
> This integration allows you to query elasticsearch instances using dsl, eql and lucene syntaxes. In addition you can fetch incidents with predefined query.
> 
> @bakatzir Can you help with the pack description?

No, the dev can curate a few sentences from the design doc usually, and tech-docs do the copy.

**Comment by iyeshaya** [comment]
@ShirleyDenkberg Let me know if this is adequate.

Elasticsearch is the distributed search and analytics engine at the heart of the Elastic Stack and where the indexing, search, and analysis magic happens. Elasticsearch offers speed and flexibility to handle data in a wide variety of use cases:

Add a search box to an app or website
Store and analyze logs, metrics, and security event data
Use machine learning to automatically model the behavior of your data in real time
Automate business workflows using Elasticsearch as a storage engine
Manage, integrate, and analyze spatial information using Elasticsearch as a geographic information system (GIS)
Store and process genetic data using Elasticsearch as a bioinformatics research tool

This pack provides an integration with the [elasticsearch API](https://www.elastic.co/guide/en/elasticsearch/reference/current/index.html)  and allows you to query elasticsearch instances using dsl, eql and lucene syntaxes. In addition you can fetch incidents with predefined query.

File: Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md [hunk 7]
```diff
@@ -1,339 +1,172 @@
...
-<h3>Setup schema mapping</h3> 
-<ol>
-<li>Navigate to <strong>Settings</strong> &gt; <strong>Integrations</strong>  &gt; <strong>Classification &amp; Mapping</strong>.</li>
-<li>Create a new mapper and edit an existing one</li>
-<li>Set <strong>Get data</strong> to <strong>Select schema</strong> and <strong>Select instance</strong> to the Elasticsearch v2 integration instance to fetch from.</li>
-<li>Map fields according to the fetched index schema.</li>
-</ol>
+Search for and analyze data in real time. 
+ Supports version 6 and later.
+This integration was integrated and tested with version xx of Elasticsearch v2
```
Comment (ShahafBenYakir) [inline Packs/Elasticsearch/Integrations/Elasticsearch_v2/README.md:3]: xx?

**Review by ShahafBenYakir — APPROVED** [review]

**Pull request merged**